[cgroups2] Cgroups v2 isolator initialization and mount setup.

DevinLeamy opened the pull request:
Updates the cgroups v2 isolator to include initialization, cleanup, update, and recovery logic.

Unlike cgroups v1 we:
- Create a new cgroup namespace during isolation, by introducing a new clone namespace flag. This implies that the contained process will only have access to cgroups in its cgroup subtree
- We only need to recover two cgroups (the domain and leaf cgroups [1]) for each container, rather than having to recover one cgroup for each controller the container used.
- We do not (yet) support nested containers.

Using the cgroups v2 isolator requires Mesos to be compiled with `--enable-cgroups-v2` and to have the cgroup2 filesystem mounted at /sys/fs/cgroup. Selecting the correct isolator version (v1 or v2) is done automatically. V2 is used if the host supports cgroups v2 and is correctly configured.

[1] The "domain" cgroup is the cgroup for a container where resource
    constraints are imposed. The "leaf" cgroup, which has the path
    <domain cgroup>/leaf, is where the container PID is put.
    Container PIDs are only put in leaf cgroups.

bmahler reviewed (commented):
commit description seems more like as:

```
[cgroups2] Implement Cgroups 2 isolator w/o nested containers and systemd.
```

and in the commit message bullets it would be good to mention the lack of systemd integration.

The domain terminology in this patch / commit message is confusing, since AFAICT all of our cgroups including the leaf ones will have cgroup.type == "domain"?

DevinLeamy commented:
Renamed "domain cgroup" to "non-leaf cgroup". Added a comment with a diagram to the Cgroups2IsolatorProcess header to illustrate the cgroup structure for a container.

DevinLeamy commented on src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.hpp:
```diff
@@ -32,6 +35,23 @@ namespace mesos {
+// container is associated with two cgroups: a non-leaf cgroup whose control
+// files are updated and a leaf cgroup where the container process lives.
+// The container pid cannot live in the non-leaf cgroup because of the cgroups
+// v2 internal process constraint.
+// Learn more here: https://docs.kernel.org/admin-guide/cgroup-v2.html#no-internal-process-constraint
```
Add no-lint.

DevinLeamy commented on src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp:
```diff
@@ -120,6 +144,569 @@ bool Cgroups2IsolatorProcess::supportsStandalone()
+  vector<Future<Nothing>> prepares;
+  foreachvalue (const Owned<Controller>& controller, controllers) {
+    // Enable controllers in the non-leaf cgroup.
+    Try<Nothing> enable =
+      cgroups2::controllers::enable(cgroup, {controller->name()});
```
Enable the controllers in the leaf controllers so that containers have the ability to manage their own resources and/or enable controllers in sub cgroups.

DevinLeamy commented on src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp:
```diff
@@ -120,6 +144,569 @@ bool Cgroups2IsolatorProcess::supportsStandalone()
+Future<Nothing> Cgroups2IsolatorProcess::___recover(
+    const ContainerID& containerId)
+{
+  // Check if:
+  // 1. The non-leaf cgroup exists                    => Fail is not found.
+  // 2. The leaf cgroup exists                        => Fail is not found.
+  // 3. Each of the requested controllers are enabled => Log and continue.
+  //
+  // Failure modes:
+  // 1. Container fails during launch.
+  //    This can happen if the launcher fails to `fork`, 'this' isolator fails
+  //    to `prepare` or `isolate`, among other reasons. Cgroups may be
+  //    improperly configured meaning there may be missing cgroups or cgroup
+  //    control files that have the wrong values.
+  // 2. Container fails on destroy.
+  //    The container fails to be destroyed so cgroups may not have been
+  //    correctly cleaned up. This can result in orphan cgroups.
+  // 3. Mesos agent is restarted with different flags.
+  //    If the agent is started with new isolators the cgroups for the existing
+  //    containers, from a previous run, won't have all the requested
+  //    controllers enabled.
```
Provide more justification for why it is okay to bail here when there is a missing cgroup or cgroups.

DevinLeamy replied:
Updated to handle invalid container states more gracefully to reduce the likelihood of bailing out.

bmahler reviewed (commented):
Local edits on commit:

```c++
diff --git a/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp b/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp
index 3784bd381..911eadb1e 100644
--- a/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp
+++ b/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp
@@ -154,43 +154,44 @@ Future<Option<ContainerLaunchInfo>> Cgroups2IsolatorProcess::prepare(
   }
 
   if (infos.contains(containerId)) {
-    return Failure(
-        "Container with id '" + stringify(containerId) + "' "
-        "has already been prepared");
+    return Failure("Container with id '" + stringify(containerId) + "'"
+                   " has already been prepared");
   }
 
   CHECK(containerConfig.container_class() != ContainerClass::DEBUG);
 
   // Create the non-leaf and leaf cgroups for the container, enable
   // controllers in the non-leaf cgroup, and `prepare` each of the controllers.
-  const string& nonLeafCgroup = cgroups2_paths::container(
+  const string nonLeafCgroup = cgroups2_paths::container(
       flags.cgroups_root, containerId);
   if (cgroups2::exists(nonLeafCgroup)) {
     return Failure("Cgroup '" + nonLeafCgroup + "' already exists");
   }
 
-  Try<Nothing> createNonLeafCgroup = cgroups2::create(nonLeafCgroup);
-  if (createNonLeafCgroup.isError()) {
+  Try<Nothing> create = cgroups2::create(nonLeafCgroup);
+  if (create.isError()) {
     return Failure("Failed to create cgroup '" + nonLeafCgroup + "': "
-                   + createNonLeafCgroup.error());
+                   + create.error());
   }
 
-  const string& leafCgroup = cgroups2_paths::container(
+  const string leafCgroup = cgroups2_paths::container(
       flags.cgroups_root, containerId, true);
   if (cgroups2::exists(leafCgroup)) {
     return Failure("Cgroup '" + leafCgroup + "' already exists");
   }
 
-  Try<Nothing> createLeafCgroup = cgroups2::create(leafCgroup);
-  if (createLeafCgroup.isError()) {
+  create = cgroups2::create(leafCgroup);
+  if (create.isError()) {
     return Failure("Failed to create cgroup '" + leafCgroup + "': "
-                   + createLeafCgroup.error());
+                   + create.error());
   }
 
-  LOG(INFO) << "Created cgroups '" << nonLeafCgroup << "' "
-            << "and '" << leafCgroup << "'";
+  LOG(INFO) << "Created cgroups '" << nonLeafCgroup << "'"
+            << " and '" << leafCgroup << "'";
+
+  infos[containerId] = Owned<Info>(
+      new Info(containerId, nonLeafCgroup, leafCgroup));
 
-  infos[containerId] = Owned<Info>(new Info(containerId, nonLeafCgroup));
   vector<Future<Nothing>> prepares;
   foreachvalue (const Owned<Controller>& controller, controllers) {
     if (controller->name() == "core") {
@@ -204,22 +205,19 @@ Future<Option<ContainerLaunchInfo>> Cgroups2IsolatorProcess::prepare(
       continue;
     }
 
-    Try<Nothing> enableInNonLeaf =
+    Try<Nothing> enable =
       cgroups2::controllers::enable(nonLeafCgroup, {controller->name()});
-    if (enableInNonLeaf.isError()) {
-      return Failure(
-          "Failed to enable controller '" + controller->name() + "' "
-          "in cgroup '" + nonLeafCgroup + "': " + enableInNonLeaf.error());
+    if (enable.isError()) {
+      return Failure("Failed to enable controller '" + controller->name() + "'"
+                     " in cgroup '" + nonLeafCgroup + "': " + enable.error());
     }
 
     // We enable controllers in the leaf cgroup to allow the container process
     // to manage their own cgroups, if they choose.
-    Try<Nothing> enableInLeaf =
-      cgroups2::controllers::enable(leafCgroup, {controller->name()});
-    if (enableInLeaf.isError()) {
-      return Failure(
-          "Failed to enable controllers in cgroup '" + nonLeafCgroup + "': "
-          + enableInLeaf.error());
+    enable = cgroups2::controllers::enable(leafCgroup, {controller->name()});
+    if (enable.isError()) {
+      return Failure("Failed to enable controllers in cgroup"
+                     " '" + nonLeafCgroup + "': " + enable.error());
     }
 
     infos[containerId]->controllers.insert(controller->name());
@@ -250,8 +248,8 @@ Future<Option<ContainerLaunchInfo>> Cgroups2IsolatorProcess::_prepare(
   }
 
   if (!errors.empty()) {
-    return Failure(
-        "Failed to prepare controllers: " + strings::join(": ", errors));
+    return Failure("Failed to prepare controllers: "
+                   + strings::join(", ", errors));
   }
 
   return update(
@@ -271,29 +269,56 @@ Future<Option<ContainerLaunchInfo>> Cgroups2IsolatorProcess::__prepare(
     const ContainerConfig& containerConfig)
 {
   // Only create cgroup mounts for containers with rootfs.
+  //
+  // TODO(bmahler): Consider adding cgroup namespace isolation for containers
+  // without a rootfs, which seems to be a useful feature?
   if (!containerConfig.has_rootfs()) {
     return None();
   }
 
   Owned<Info> info = cgroupInfo(containerId);
   if (!info.get()) {
-    return Failure(
-        "Failed to get cgroup for container "
-        "'" + stringify(containerId) + "'");
+    return Failure("Failed to get cgroup for container"
+                   " '" + stringify(containerId) + "'");
   }
 
   ContainerLaunchInfo launchInfo;
+
   // Create a new cgroup namespace. The child process will only be able to
   // see the cgroups that are in its cgroup subtree.
   launchInfo.add_clone_namespaces(CLONE_NEWCGROUP);
 
   // Create a new mount namespace and mount the root cgroup at /sys/fs/cgroup.
+  // TODO(bmahler): Is this the right way to mount?
   launchInfo.add_clone_namespaces(CLONE_NEWNS);
   *launchInfo.add_mounts() = protobuf::slave::createContainerMount(
-      cgroups2::path(info->cgroup),
+      cgroups2::path(info->cgroup_leaf) ,
       path::join(containerConfig.rootfs(), "/sys/fs/cgroup"),
       MS_BIND | MS_REC);
 
+  // TODO(qianzhang): This is a hack to pass the container-specific cgroups
+  // mounts and the symbolic links to the command executor to do for the
+  // command task. The reasons that we do it in this way are:
+  //   1. We need to ensure the container-specific cgroups mounts are done
+  //      only in the command task's mount namespace but not in the command
+  //      executor's mount namespace.
+  //   2. Even it's acceptable to do the container-specific cgroups mounts
+  //      in the command executor's mount namespace and the command task
+  //      inherit them from there (i.e., here we just return `launchInfo`
+  //      rather than passing it via `--task_launch_info`), the container
+  //      specific cgroups mounts will be hidden by the `sysfs` mounts done in
+  //      `mountSpecialFilesystems()` when the command executor launches the
+  //      command task.
+  if (containerConfig.has_task_info()) {
+    ContainerLaunchInfo _launchInfo;
+
+    _launchInfo.mutable_command()->add_arguments(
+        "--task_launch_info=" +
+        stringify(JSON::protobuf(launchInfo)));
+
+    return _launchInfo;
+  }
+
   return launchInfo;
 }
 
@@ -302,11 +327,13 @@ Future<Nothing> Cgroups2IsolatorProcess::recover(
     const vector<ContainerState>& states,
     const hashset<ContainerID>& orphans)
 {
+  // Recover containers from checkpointed data:
   vector<Future<Nothing>> recovers;
   foreach (const ContainerState& state, states) {
     recovers.push_back(___recover(state.container_id()));
   }
 
+  // Then recover containers we find in the cgroups hierarchy:
   return await(recovers)
     .then(defer(
         PID<Cgroups2IsolatorProcess>(this),
@@ -327,9 +354,9 @@ Future<Nothing> Cgroups2IsolatorProcess::_recover(
     }
   }
 
-  if (errors.size() > 0) {
-    return Failure("Failed to recover active containers: " +
-                   strings::join(": ", errors));
+  if (!errors.empty()) {
+    return Failure("Failed to recover active containers: "
+                   + strings::join(", ", errors));
   }
 
   hashset<ContainerID> knownOrphans;
@@ -349,9 +376,9 @@ Future<Nothing> Cgroups2IsolatorProcess::_recover(
     Option<ContainerID> containerId = cgroups2_paths::containerId(
         flags.cgroups_root, cgroup);
     if (containerId.isNone()) {
-        LOG(INFO) << "Cgroup '" << cgroup << "' does not correspond to a "
-                  << "container id and will not be recovered";
-        continue;
+      LOG(INFO) << "Cgroup '" << cgroup << "' does not correspond to a"
+                << " container id and will not be recovered";
+      continue;
     }
 
     if (infos.contains(*containerId)) {
@@ -392,14 +419,14 @@ Future<Nothing> Cgroups2IsolatorProcess::__recover(
       errors.push_back(future.isFailed() ? future.failure() : "discarded");
     }
   }
-  if (errors.size() > 0) {
-    return Failure(
-        "Failed to recover orphan containers: " +
-        strings::join(": ", errors));
+  if (!errors.empty()) {
+    return Failure("Failed to recover orphan containers: "
+                   + strings::join(", ", errors));
   }
 
   // Known orphan cgroups will be destroyed by the containerizer using
-  // the normal cleanup path.
+  // the normal cleanup path, but for unknown orphans we need to clean
+  // them up here:
   foreach (const ContainerID& containerId, unknownOrphans) {
     LOG(INFO) << "Cleaning up unknown orphaned container " << containerId;
     cleanup(containerId);
@@ -415,74 +442,72 @@ Future<Nothing> Cgroups2IsolatorProcess::___recover(
   // Remark and handle invalid container states and recover enabled controllers.
   //
   // Invalid container states:
-  // 1. Missing non-leaf cgroup                       => Log and create cgroup
-  // 2. Missing leaf cgroup                           => Log and create cgroup
-  // 3. Each of the requested controllers are enabled => Log
+  // 1. Missing non-leaf cgroup            => Log and create cgroup
+  // 2. Missing leaf cgroup                => Log and create cgroup
+  // 3. Some controllers are not enabled   => Log
   //
   // Failure modes that can lead to an invalid container state:
-  // 1. Container fails during launch.
+  //
+  // 1. Mesos agent is restarted during launch.
   //    This can happen if the launcher fails to `fork`, 'this' isolator fails
   //    to `prepare` or `isolate`, among other reasons. Cgroups may be
   //    improperly configured meaning there may be missing cgroups or cgroup
   //    control files that have the wrong values.
-  // 2. Container fails on destroy.
+  // 2. Mesos agent is restarted during destroy.
   //    The container fails to be destroyed so cgroups may not have been
   //    cleaned up correctly. This can result in orphan cgroups.
   // 3. Mesos agent is restarted with different flags.
   //    If the agent is started with new isolators the cgroups for the existing
   //    containers, from a previous run, won't have all the requested
   //    controllers enabled.
-
+  //
+  // If a container is missing a cgroup, we create the missing cgroup. This
+  // is done exclusively so that the container can be cleanup()ed up by 'this'
+  // isolator and destroy()ed by the launcher like other containers.
+  // The alternative would be to break the invariant that each container has
+  // a leaf and non-leaf cgroup but that requires more special-case handling.
   const string nonLeafCgroup =
-    cgroups2_paths::container(flags.cgroups_root, containerId);
+    cgroups2_paths::container(flags.cgroups_root, containerId, false);
   const string leafCgroup =
     cgroups2_paths::container(flags.cgroups_root, containerId, true);
 
-  // If a container is missing a cgroup, we create the missing cgroup. This
-  // is done exclusively so that the container can be `cleanup`ed up by 'this'
-  // isolator and `destroy`ed by the launcher like other containers.
-  // The alternative would be to break the invariant that each container has
-  // a leaf and non-leaf cgroup but that requires more special-case handling.
   if (!cgroups2::exists(nonLeafCgroup)) {
-    LOG(WARNING) << "Container '" << stringify(containerId) << "' "
-                 << "is missing the cgroup '" << nonLeafCgroup << "'; "
-                 << "creating missing cgroup";
+    LOG(WARNING) << "Container '" << stringify(containerId) << "'"
+                 << " is missing the cgroup '" << nonLeafCgroup << "';"
+                 << " creating missing cgroup";
 
-    Try<Nothing> createNonLeaf = cgroups2::create(nonLeafCgroup);
-    if (createNonLeaf.isError()) {
+    Try<Nothing> create = cgroups2::create(nonLeafCgroup);
+    if (create.isError()) {
       return Failure("Failed to create cgroup '" + nonLeafCgroup + "': "
-                     + createNonLeaf.error());
+                     + create.error());
     }
   }
 
   if (!cgroups2::exists(leafCgroup)) {
-    LOG(WARNING) << "Container '" << stringify(containerId) << "' "
-                 << "is missing the cgroup '" << leafCgroup << "'; "
-                 << "creating missing cgroup";
+    LOG(WARNING) << "Container '" << stringify(containerId) << "'"
+                 << " is missing the cgroup '" << leafCgroup << "';"
+                 << " creating missing cgroup";
 
-    Try<Nothing> createLeaf = cgroups2::create(leafCgroup);
-    if (createLeaf.isError()) {
+    Try<Nothing> create = cgroups2::create(leafCgroup);
+    if (create.isError()) {
       return Failure("Failed to create cgroup '" + leafCgroup + "': "
-                     + createLeaf.error());
+                     + create.error());
     }
   }
 
-  const Try<set<string>> enabledControllers = cgroups2::controllers::enabled(
-      nonLeafCgroup);
-  if (enabledControllers.isError()) {
-    return Failure(
-        "Failed to get the enabled controllers for container "
-        "'" + stringify(containerId) + "': " + enabledControllers.error());
+  Try<set<string>> enabled = cgroups2::controllers::enabled(nonLeafCgroup);
+  if (enabled.isError()) {
+    return Failure("Failed to get the enabled controllers for container"
+                   " '" + stringify(containerId) + "': " + enabled.error());
   }
 
   vector<Future<Nothing>> recovers;
   hashset<string> recoveredControllers;
   foreachvalue (const Owned<Controller>& controller, controllers) {
-    if (enabledControllers->count(controller->name()) == 0) {
+    if (enabled->count(controller->name()) == 0) {
       // Controller is expected to be enabled but isn't.
-      LOG(WARNING) << "Controller '" << controller->name() << "' "
-                   << "is not enabled for container "
-                   << "'" << stringify(containerId) << "'";
+      LOG(WARNING) << "Controller '" << controller->name() << "' is not enabled"
+                   << " for container '" << stringify(containerId) << "'";
 
       continue;
     }
@@ -515,13 +540,16 @@ Future<Nothing> Cgroups2IsolatorProcess::____recover(
     }
   }
 
-  if (errors.size() > 0) {
-    return Failure("Failed to recover controllers: " +
-                   strings::join(": ", errors));
+  if (!errors.empty()) {
+    return Failure("Failed to recover controllers: "
+                   + strings::join(", ", errors));
   }
 
   infos[containerId] = Owned<Info>(new Info(
-      containerId, cgroups2_paths::container(flags.cgroups_root, containerId)));
+      containerId,
+      cgroups2_paths::container(flags.cgroups_root, containerId, false),
+      cgroups2_paths::container(flags.cgroups_root, containerId, true)));
+
   infos[containerId]->controllers = recoveredControllers;
 
   return Nothing();
@@ -567,8 +595,8 @@ Future<Nothing> Cgroups2IsolatorProcess::_isolate(
   }
 
   if (!errors.empty()) {
-    return Failure(
-        "Failed to prepare controllers: " + strings::join(": ", errors));
+    return Failure("Failed to prepare controllers: "
+                   + strings::join(", ", errors));
   }
 
   Owned<Info> info = cgroupInfo(containerId);
@@ -577,15 +605,11 @@ Future<Nothing> Cgroups2IsolatorProcess::_isolate(
         "Failed to find cgroup for container '" + stringify(containerId) + "'");
   }
 
-  // Move the process into its leaf cgroup. Since info->cgroup is the non-leaf
-  // cgroup, we get the corresponding leaf cgroup by appending /leaf. It's
-  // necessary to move into the leaf cgroup because of the internal process
-  // constraint.
-  Try<Nothing> assign = cgroups2::assign(path::join(info->cgroup, "leaf"), pid);
+  // Move the process into its leaf cgroup.
+  Try<Nothing> assign = cgroups2::assign(info->cgroup_leaf, pid);
   if (assign.isError()) {
-    return Failure(
-        "Failed to assign container '" + stringify(containerId) + "' "
-        "to cgroup '" + info->cgroup + "': " + assign.error());
+    return Failure("Failed to assign container '" + stringify(containerId) + "'"
+                   " to cgroup '" + info->cgroup + "': " + assign.error());
   }
 
   return Nothing();
@@ -598,14 +622,13 @@ Future<Nothing> Cgroups2IsolatorProcess::update(
     const google::protobuf::Map<string, Value::Scalar>& resourceLimits)
 {
   if (!infos.contains(containerId)) {
-    return Failure(
-        "Container with id '" + stringify(containerId) + "' is unknown");
+    return Failure("Unknown container '" + stringify(containerId) + "'");
   }
 
   vector<Future<Nothing>> updates;
 
-  LOG(INFO) << "Updating controllers for cgroup '"
-          << infos[containerId]->cgroup << "'";
+  LOG(INFO) << "Updating controllers for cgroup"
+            << " '" << infos[containerId]->cgroup << "'";
 
   foreachvalue (const Owned<Controller>& controller, controllers) {
     if (infos[containerId]->controllers.contains(controller->name())) {
@@ -636,8 +659,8 @@ Future<Nothing> Cgroups2IsolatorProcess::_update(
   }
 
   if (!errors.empty()) {
-    return Failure(
-        "Failed to update controllers: " + strings::join(": ", errors));
+    return Failure("Failed to update controllers: "
+                   + strings::join(", ", errors));
   }
 
   return Nothing();
@@ -716,10 +739,9 @@ Future<Nothing> Cgroups2IsolatorProcess::_cleanup(
     }
   }
 
-  if (errors.size() > 0) {
-    return Failure(
-        "Failed to cleanup subsystems: " +
-        strings::join(";", errors));
+  if (!errors.empty()) {
+    return Failure("Failed to cleanup subsystems: "
+                   + strings::join(", ", errors));
   }
 
   if (cgroups2::exists(infos[containerId]->cgroup)) {
diff --git a/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.hpp b/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.hpp
index 0b4d90596..211b9a1df 100644
--- a/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.hpp
+++ b/src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.hpp
@@ -39,19 +39,20 @@ namespace slave {
 //
 // Manages the cgroup v2 controllers that are used by containers. Each
 // container is associated with two cgroups: a non-leaf cgroup whose control
-// files are updated and a leaf cgroup where the container process lives.
+// files are updated and a leaf cgroup where the container's processes lives.
 // The container pid cannot live in the non-leaf cgroup because of the cgroups
-// v2 internal process constraint.
-// Learn more here: https://docs.kernel.org/admin-guide/cgroup-v2.html#no-internal-process-constraint // NOLINT
+// v2 internal process constraint:
 //
-// Example:
+// https://docs.kernel.org/admin-guide/cgroup-v2.html#no-internal-process-constraint // NOLINT
+//
+// Example cgroups:
 //     containerA                       non-leaf cgroup
 //     /      \                         /            \
 // processes  containerB           leaf cgroup   non-leaf child cgroup
 //             |                                      |
 //            processes                          leaf-cgroup
 //
-// Note: Nested containers are not yet supported.
+// TODO(dleamy): Nested containers are not yet supported.
 class Cgroups2IsolatorProcess : public MesosIsolatorProcess
 {
 public:
@@ -88,8 +89,10 @@ public:
 private:
   struct Info
   {
-    Info(const ContainerID& _containerId, const std::string& _cgroup)
-      : containerId(_containerId), cgroup(_cgroup) {}
+    Info(const ContainerID& containerId,
+         const std::string& cgroup,
+         const std::string& cgroup_leaf)
+      : containerId(containerId), cgroup(cgroup), cgroup_leaf(cgroup_leaf) {}
 
     const ContainerID containerId;
 
@@ -97,6 +100,7 @@ private:
     // updated to set resource constraints on this and descendant
     // containers. Processes should not be assigned to this cgroup.
     const std::string cgroup;
+    const std::string cgroup_leaf;
 
     // Names of the controllers which are prepared for the container.
     hashset<std::string> controllers;
```
